ArduCopter/toy_mode.cpp

```diff
@@ -1080,4 +1096,70 @@ void ToyMode::arm_check_compass(void)
     }
 }
 
+/*
+  handle take of cmd
+ */
+void ToyMode::handle_takeoff_cmd(bool takeoff_btn,bool takeoff_cancel)
+{
+    const uint32_t takeoff_delay_ms = 3000;
+    
+    //initiate takeoff cmd
+    if (takeoff_state == TAKEOFF_ON_GROUND && !copter.motors->armed() && takeoff_btn) {
+        takeoff_state = TAKEOFF_INITIATE;
+        takeoff_last_press_ms = AP_HAL::millis();
+    }
+    
+    switch (takeoff_state) {
+    case TAKEOFF_INITIATE:
+        
+        if (!copter.motors->armed()) {
+            action_arm();
+        }
+        
+        if (!copter.init_arm_motors(true) && (copter.control_mode == LOITER)) {
+            /*
+                support auto-switching to ALT_HOLD
+            */
+            if (set_and_remember_mode(get_non_gps_mode(), MODE_REASON_TMODE)) {
+                action_arm();
+                GCS_MAVLINK::send_statustext_all(MAV_SEVERITY_INFO, "Tmode: indoor update arm");
+            }
+        }
+        
+        if (copter.motors->armed()) {
+            //delay the take off to accomodate for the initial arming of the copter
+            if (!((AP_HAL::millis() - takeoff_last_press_ms) < takeoff_delay_ms)) {
+                takeoff_state = TAKEOFF_ARMED;
+            }
+        }
+        break;
+    case TAKEOFF_ARMED:
```

Comment:
Possibly rename state to something describing what's being attempted in this state - `TAKEOFF_DO_ASCEND` or `TAKEOFF_DO_CLIMB`?